STM32 SPI DMA bug fixes and enhancements

lowlander opened the pull request:
This PR has several bug fixes and enhancements for the STM32 SPI driver when using DMA

1) When using GPIO CS the CS select was enabled after the DMA started the transfer, resulting in the first few bits being transferred while CS was still disabled.

2) when doing RX or TX only the driver hung and never could finish the transfer

3) The driver used a busy wait loop to wait for DMA to finish, using a lot of CPU for large (20ms+ transfers). Now the DMA ready is signaled with a semaphore.

4) A small bug fix to the DMA driver that caused a delay in finishing the SPI transfer

lowlander committed:
drivers: spi_ll_stm32: refactor DMA support

- Fix GPIO CS timing when using DMA. When using GPIO CS the
  CS select was enabled after the DMA started the transfer,
  resulting in the first few bits being transfered while
  CS was still disabled.

- Fix TX or RX only DMA transfers. When only a RX or only
  a TX transfer was requested the DMA never finished.

  For the RX only cause the size on the transfer was
  calculated by taking the TX buffer length (0), this
  caused problems.

  For the TX only transfer the RX buffer was set to NULL,
  this caused the DMA to acctually writing data to the
  adress 0x00000000. By using the dummy destination buffer
  it now only writes to valid memory.

- Add semaphore to signal that DMA is ready, instead of
  just busy waiting.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lochej reviewed (approved):
LGTM as long as you have been able to test that the first bits are correctly transmitted.

lowlander commented:
> LGTM as long as you have been able to test that the first bits are correctly transmitted.

It looks good on my scope, this is a SPI OLED display, hence the big transfer. (which also explains why I need DMA);

![spi_oled_good](https://user-images.githubusercontent.com/1266398/89016936-d5a19100-d319-11ea-93ae-7d8c4d946375.png)

aurel32 commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -716,6 +717,7 @@ static int transceive_dma(struct device *dev,
 
 		/* once SPI is no more busy, wait for DMA transfer end */
 		while (spi_stm32_dma_transfer_ongoing(data) == 1) {
+			k_sem_take(&data->dma_status_sem, K_MSEC(100));
```
A semaphore is definitely better than the busy loop, as it allows the CPU to execute other tasks or sleep. A busy loop kind of defeat the purpose of DMA...

Now I have two points:
- The underlying bug is that the `transfer_complete` bool in struct stream is not declared volatile, so the compiler optimizes the loop as a if test and an endless loop. I am afraid that even after your change, the bug is still there. Now that there is a semaphore, I believe that the `transfer_complete` booleans  are redundant an not needed anymore.
- You used a timeout in `k_sem_take`. I would suggest either to use `K_FOREVER` or to properly handle the timeout by checking the value returned this function

lowlander replied:
If the DMA flags are just flags and not volatile (or atomic, or some sort of MB) that seems broken by design :-/ I didn't run into that problem yet, will take a look if it also can cause me problems. 

The timeout was thought as preparation for a real timeout so it will be possible to recover from when "something" goes badly goes wrong (better one broken transfer than a system that hangs for ever). For now it shouldn't matter that the k_sem_take wakes up every 100ms, it will check if the dma is still ongoing and go back to sleep.  But if it is a show-stopper I can make a K_FOREVER out of it.

lowlander replied:
> * The underlying bug is that the `transfer_complete` bool in struct stream is not declared volatile, so the compiler optimizes the loop as a if test and an endless loop. I am afraid that even after your change, the bug is still there. Now that there is a semaphore, I believe that the `transfer_complete` booleans  are redundant an not needed anymore.

I have one semaphore for signaling RX and TX state (the two flags still say if RX and TX are both ready), maybe removing those flags and using 2 semaphores to signal TX and RX are done. One could first take TX and than take RX. 

PS: the k_sem_take in the loop might fix your optimizing problem because the k_sem_take code surely has some kind of memory barrier code in it so the compiler can't optimize the loop away anymore.

aurel32 replied:
> > * The underlying bug is that the `transfer_complete` bool in struct stream is not declared volatile, so the compiler optimizes the loop as a if test and an endless loop. I am afraid that even after your change, the bug is still there. Now that there is a semaphore, I believe that the `transfer_complete` booleans  are redundant an not needed anymore.
> 
> I have one semaphore for signaling RX and TX state (the two flags still say if RX and TX are both ready), maybe removing those flags and using 2 semaphores to signal TX and RX are done. One could first take TX and than take RX.

Yes that indeed looks like the way to go, right now the semaphores and the booleans provide a redundant information.

> PS: the k_sem_take in the loop might fix your optimizing problem because the k_sem_take code surely has some kind of memory barrier code in it so the compiler can't optimize the loop away anymore.

Yes that's a good point. Actually given that k_sem_take is defined in a different file and LTO is not enabled, the compiler can't do any optimization anymore.

lowlander replied:
Just realized that one sem should be enough it just has to be given twice (for TX and RX) and taken twice, since the code is only interested in knowning that both RX and TX are done. But on the other hand that does mean it must be absolutly certain that both RX and TX DMA complete signals come only 1 time, so maybe a TX and RX sem is more robust.

erwango commented:
@aurel32 are you fine with latest status ?

lowlander commented:
@erwango @aurel32  I would like to try with the RX and TX semaphore first, but won't get to it until Friday.

lowlander commented:
@lochej some how github doesn't let me request a re-review by you, but could you take a look at the semaphore code?

lowlander commented:
@aurel32 could you take a look at the sem code ?

aurel32 reviewed (changes requested):
Otherwise it looks good to me, but I haven't been able to try on real hardware yet.

lowlander commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -718,6 +718,10 @@ static int transceive_dma(struct device *dev,
+		/* wait until TX buffer is really empty */
+		while (LL_I2S_IsActiveFlag_TXE(spi) == 0) {
+		}
+
```
This looks wrong, let me take a look at that.

lochej replied:
I2S looks wrong in an SPI driver :upside_down_face:

lowlander replied:
Well it seems LL_I2S_IsActiveFlag_TXE and LL_SPI_IsActiveFlag_TXE are both available on F4, I think I simply didn't notice the _I2S_ in the name. Luckily on F4 the _i2S_ version just calls the _SPI_ version, so it generated the same code.

lowlander commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -159,19 +159,27 @@ static int spi_stm32_dma_rx_load(struct device *dev, uint8_t *buf, size_t len)
-	blk_cfg.dest_address = (buf != NULL) ? (uint32_t)buf : (uint32_t)NULL;
+	if (buf == NULL) {
+		/* if rx buff is null, then write data to dummy address. */
+		blk_cfg.dest_address = (uint32_t)&nop_tx;
```
Hmmm yeah wanted to do that but first wanted to try if it worked, and of course forgot.

lochej reviewed (approved):
LGTM code wise. Semaphore usage is OK to me, given and taken appropriately. I'm unable to test now because I lack a correct test platform but I'll leave that to other reviewers.

nandojve commented:
Question: Is DMA working on L053 too? If yes, can someone help with UART-1 and SPI-1 configs?

lowlander commented:
> Question: Is DMA working on L053 too? If yes, can someone help with UART-1 and SPI-1 configs?

Sorry @nandojve I don't have L053 hardware, so I can not test that.

lochej commented:
> Question: Is DMA working on L053 too? If yes, can someone help with UART-1 and SPI-1 configs?

I could test on my nucleo_l073rz with a correct test setup and DT configuration if you need help.

lowlander commented:
I have the feeling there is something wrong with the STM32 DMA in general, somehow it looks like channel and streams are mixed up in some places, the Zephyr's MACRO-FU is not really helpful in understanding the code. 

When I enable logging it endlessly complains about "Unknown interrupt happened". In the dma code there seems to be some code parts that can pass the channel nr to the isr (that code seems unused) but when i add it I can print out the channel nr in the DMA isr and it complains about channel 4 (which I don't use), there is a stream 4 though. So maybe the naming is wrong, and it the variable should be named stream? 

All very confusing, and DT docu isn't really helping either :-/

@lochej what is your DT setup ?

lochej commented:
> I have the feeling there is something wrong with the STM32 DMA in general, somehow it looks like channel and streams are mixed up in some places, the Zephyr's MACRO-FU is not really helpful in understanding the code.
> 
> When I enable logging it endlessly complains about "Unknown interrupt happened". In the dma code there seems to be some code parts that can pass the channel nr to the isr (that code seems unused) but when i add it I can print out the channel nr in the DMA isr and it complains about channel 4 (which I don't use), there is a stream 4 though. So maybe the naming is wrong, and it the variable should be named stream?
> 
> All very confusing, and DT docu isn't really helping either :-/
>

From looking at the bindibgs for DMA-STM32, streams and channel are determined from dmav1 or dmav2. (i.e., dmav1's stream number is dmav2's channel number).

[Dma bindings](https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/bindings/dma/st%2Cstm32-dma.yaml)

> @lochej what is your DT setup ?

Don't have any yet, but I could build one to test this out.

aurel32 commented:
@lowlander Thanks for the code changes, code wise it looks all good. I gave it a try for the communication between an SX1276 module and a STM32L433 and it doesn't work. But it didn't work before this PR either. I'll try to debug that in the next days.

lowlander commented:
> @lowlander Thanks for the code changes, code wise it looks all good. I gave it a try for the communication between an SX1276 module and a STM32L433 and it doesn't work. But it didn't work before this PR either. I'll try to debug that in the next days.

@aurel32 any idea what goes wrong? Is there any difference between the L and F series when it comes to DMA/SPI ?

lowlander commented:
> 
> [Dma bindings](https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/bindings/dma/st%2Cstm32-dma.yaml)
> 

@lochej the documentation and naming doesn't make it easier, from that doc;
```
  1. channel: the dma stream from 0 in V1 or 1 in V2 to <dma-requests>
  2. slot: DMA periph request ID in DMA_V2 or DMAMUX, dma request in V1
```

So the first number is the channel (but really it is the stream) and the second the "slot" is the ID (what the ST docu calls the channel).

Also from that doc

```
For the client part, example for stm32f411 DMA V1 on DMA2 instance
    Tx using stream 5 on channel 3 (stream 2 on channel 2 is also possible)
    Rx using stream 2 on channel 3 (stream 0 on channel 3 is also possible)
    spi1 {
     dmas = <&dma2 5 3 0x28440 0x03>,
           <&dma2 2 3 0x28480 0x03>;
     dma-names = "tx", "rx";
     };
```
So the channel is the stream and the slot is the channel, not confusing at all. 

And when I look at the `dma_callback(struct device* dev, void* arg, uint32_t channel, int status)` in the spi driver, that is the stream not the channel, but it is called channel. 

So channel, slot, id, stream, and request all seem to be used at random, this makes it very hard to understand the code. 

But it also means my DT should be correct. :-)

nandojve commented:
> > Question: Is DMA working on L053 too? If yes, can someone help with UART-1 and SPI-1 configs?
> 
> I could test on my nucleo_l073rz with a correct test setup and DT configuration if you need help.

I tried this, without success. It builds, and board run but at first spi communication try it hangs:

**nucleo_l053r8**

config SPI_STM32_DMA
	default **y**
	depends on SPI

config SPI_STM32_INTERRUPT
	default **n**
	depends on SPI

config SPI_STM32_USE_HW_SS
	default **n**
	depends on SPI

```
&spi1 {
       /* Not sure if below is right too */
	dmas = <&dma1 3 1 0x20240 0x04>,
	       <&dma1 2 1 0x20480 0x04>;
	dma-names = "tx", "rx";
};
&dma1 {
	status = "okay";
};
```

> So channel, slot, id, stream, and request all seem to be used at random, this makes it very hard to understand the code.

What mess ^^!!

lochej commented:
@nandojve AN2548 shows that stm32L series are using DMA V2.
Make sure your DT uses V2 definition. On the other hand, the DMA code seems to be designed for V1 with stream definitions. I don't know if information can help but DMA handling might need to be refactored to differ based on CPU type (STM32L series, STM32F4 series...).

[STM32F0/F1/F3/G0Lx DMA description V2](https://www.google.com/url?q=https://www.st.com/resource/en/application_note/cd00160362-using-the-stm32f0-f1-f3-g0-lx-series-dma-controller-stmicroelectronics.pdf&sa=U&ved=2ahUKEwjWkMGvuaDrAhXh4IUKHZ0oAoYQFjAAegQIChAB&usg=AOvVaw0JHSQZwXxpQn30D65w1IWV)

lowlander commented:
> @lowlander @nandojve AN2548 shows that stm32L series are using DMA V2.
> Make sure your DT uses V2 definition. On the other hand, the DMA code seems to be designed for V1 with stream definitions. I don't know if information can help but DMA handling might need to be refactored to differ based on CPU type (STM32L series, STM32F4 series...).
> 
> [STM32F0/F1/F3/G0Lx DMA description V2](https://www.google.com/url?q=https://www.st.com/resource/en/application_note/cd00160362-using-the-stm32f0-f1-f3-g0-lx-series-dma-controller-stmicroelectronics.pdf&sa=U&ved=2ahUKEwjWkMGvuaDrAhXh4IUKHZ0oAoYQFjAAegQIChAB&usg=AOvVaw0JHSQZwXxpQn30D65w1IWV)

@lochej I am really considering rewriting the whole STM32 DMA driver, There are several things I dislike about it, like the use of kmalloc, integer flags shared between isr and other code, weird function pointer setup, weird IRQ handling.

nandojve commented:
> @nandojve could you give the update a try (in combination with #27723 ) , noticed that when logging is on the highest level it hung by me too, that seems to be a race condition. I changed the init order a bit and make sure both RX and TX DMA are enabled (if there is no buffer it uses the dummy buffer, so they should always be enabled)

Hi @lowlander, not sure if you had time to look at NVIC from L0x. It seems that channels are sharing interrupt lines.
| IRQ | Acronym | Desc |
| --- | --- | --- |
| 9 | DMA1_Channel1 | DMA1 Channel1 global interrupt |
| 10 | DMA1_Channel[3:2] | DMA1 Channel2 and 3 interrupts |
| 11 | DMA1_Channel[7:4] | DMA1 Channel4 to 7 interrupts |

I got below error after enable DMA with #27263 and #27723:
```
[147/150] Generating isr_tables.c
FAILED: zephyr/isr_tables.c 
cd /home/gfbudke/zephyros/mainline/zephyr/build/zephyr && /home/gfbudke/zephyr-sdk-0.11.3/arm-zephyr-eabi/bin/arm-zephyr-eabi-objcopy --input-target=elf32-littlearm --output-target=binary --only-section=.intList /home/gfbudke/zephyros/mainline/zephyr/build/zephyr/zephyr_prebuilt.elf isrList.bin && /usr/bin/python3.7 /home/gfbudke/zephyros/mainline/zephyr/arch/common/gen_isr_tables.py --output-source isr_tables.c --kernel /home/gfbudke/zephyros/mainline/zephyr/build/zephyr/zephyr_prebuilt.elf --intlist isrList.bin   --sw-isr-table --vector-table
gen_isr_tables.py: error: multiple registrations at table_index 11 for irq 11 (0xb)
Existing handler 0x80073cf, new handler 0x80073c5
Has IRQ_CONNECT or IRQ_DIRECT_CONNECT accidentally been invoked on the same irq multiple times?
```
I believe it is related to https://github.com/zephyrproject-rtos/zephyr/pull/27723/commits/a0cfe27b84a8f0858f7631c64bf6f82e7b2e45d3. I do prefer your changes but we need some alternative for devices with shared lines.

My guess is L0 devices will share DMA int
@erwango  ^^ ?

lowlander commented:
@nandojve first of all thanks for testing, and yes it seems L0 has only 3 IRQ's for the 7 channels, I will update the code to deal with that, since it is known at compile time it should be doable.  And at the end of the week I also should have hardware to test it myself (although this compile error I could/should have found without hardware).

aurel32 commented:
> @aurel32 could you also give it a retry in combination with #27723 to see if it works on a L4 ?

Sorry I have been busy doing another setup using a nucleo l432kc board and an sx1276 instead of the final board where things are a bit more difficult to debug. I am also addressing the sx1276 directly instead of using the Zephyr driver. Therefore I haven't tested #27723 yet.

> > @lowlander Thanks for the code changes, code wise it looks all good. I gave it a try for the communication between an SX1276 module and a STM32L433 and it doesn't work. But it didn't work before this PR either. I'll try to debug that in the next days.
> 
> @aurel32 any idea what goes wrong? Is there any difference between the L and F series when it comes to DMA/SPI ?
> 
> update/edit: I noticed the F4 is V1 and L4 V2 of the driver, and the channel offsets are different, maybe that is where things go wrong, maybe you have the wrong channel/stream setup in your DT ?

I have progress a bit on that. Initially without this PR, I had this kind of error:
```
[00:00:00.206,000] <wrn> dma_stm32: dest_buffer address is null.
[00:00:00.206,000] <err> dma_stm32: dma stream 1 is busy.
[00:00:00.206,000] <err> main: TX failed: -16
```

The 3 first commits of this PR make things to work. However the last one (drivers: spi_ll_stm32: use semaphores to prevent busy wait loop) breaks things again. It breaks when `spi_write()` is called with a transaction of 2 buffers, while it works fine with 1) `spi_write()`and a single buffer or 2) `spi_transceive()` and multiple buffers.

I haven't debugged further yet.

aurel32 commented:
Just looking at the code, it seems that before we waited for either of the two transfers to stop:
```
static bool spi_stm32_dma_transfer_ongoing(struct spi_stm32_data *data)
{
       return ((data->dma_tx.transfer_complete != true)
               && (data->dma_rx.transfer_complete != true));
}

/* once SPI is no more busy, wait for DMA transfer end */
while (spi_stm32_dma_transfer_ongoing(data) == 1) {
}
```

That way they could be reloaded independently if they do not have the same length. With your change we expect both transfers to finish at the same time.

aurel32 commented:
@lowlander I wonder in which conditions the third commit "drivers: spi_ll_stm32: make sure DMA is done sending" commit is necessary and if there is another way to achieve the same result.

Without that commit, it seems that all the work that is done in the `for (; ;)` loop of the `transceive_dma` function can be done directly in the DMA interrupt, as there are all short functions without busy loop. The semaphore can then be used to signal that all the buffers have been RXed or TXed. That would actually make sense to run the SPI transfer as fast as possible without waiting between buffers depending on the priority of the thread calling `spi_write/read/transceive`.

lowlander commented:
@aurel32 From the logic there should not be any difference from the old driver, that waited for the busy to go away (which pretty much means TX is ready) and than looped on `spi_stm32_dma_transfer_ongoing`  which is the same as waiting for the two semaphores in a row. So I am not sure I understand why it works without the last commit and doesn't work with the last commit. 

One of the big problems I had with the old driver was that a RX only did not work because it seems it needs a TX to generate the clock signal. This also means that TX-len may not be smaller than RX-len, else the RX DMA will never finish. On the other hand the RX-len can be smaller for a single transfer, but when having multiple buffers part of the RX will be lost.

I think your right that the better way to handle all corner cases is to reload the DMA from the ISR, because there it is known if it is RX or TX that is done.  But I have the feeling that will be quite some work.

aurel32 commented:
> @aurel32 From the logic there should not be any difference from the old driver, that waited for the busy to go away (which pretty much means TX is ready) and than looped on `spi_stm32_dma_transfer_ongoing` which is the same as waiting for the two semaphores in a row. 

No it is not the same, as with `spi_stm32_dma_transfer_ongoing` the loop stops as soon as RX **or** TX **or** both transfers are complete. With the semaphore, the code waits for TX **and** RX transfers to finish. 

> So I am not sure I understand why it works without the last commit and doesn't work with the last commit.
 
> One of the big problems I had with the old driver was that a RX only did not work because it seems it needs a TX to generate the clock signal. This also means that TX-len may not be smaller than RX-len, else the RX DMA will never finish. On the other hand the RX-len can be smaller for a single transfer, but when having multiple buffers part of the RX will be lost.

Yes, I agree that the total TX and RX len should be the same, however I think the current implementation allows the buffers to be split differently. For example for a 5-byte transfer, TX could use 2 buffers of 2 and 3 bytes and RX a single buffer of 5 bytes.
 
> I think your right that the better way to handle all corner cases is to reload the DMA from the ISR, because there it is known if it is RX or TX that is done. But I have the feeling that will be quite some work.

i'll give it a try.

lowlander commented:
> No it is not the same, as with `spi_stm32_dma_transfer_ongoing` the loop stops as soon as RX **or** TX **or** both transfers are complete. With the semaphore, the code waits for TX **and** RX transfers to finish.

Negative boolean logic gets me every time, combined with a name "ongoing" that looks at RX and TX, I misunderstood thinking it was RX and TX ready instead of RX or TX ready :-/

PS: are you really going to give the IRQ stuff a try, not that we both write something.

aurel32 commented:
> PS: are you really going to give the IRQ stuff a try, not that we both write something.

I gave it a first try and it works well in my test case: https://github.com/aurel32/zephyr/commits/stm32-spi-dma

The code still use boolean flags instead of semaphores, it still has to be converted the same way you already did with your version. I am also not sure about the condition to check for the end of the transfer, i have chosen to test for the SPI being busy as before and the TX empty flag like in your "drivers: spi_ll_stm32: make sure DMA is done sending" commit. Note that this is done only at the end of all the whole transfer instead of each buffer, so I expect also some performance improvements when multiple buffers are used.

PS: I won't work on this before tomorrow night CEST, feel free to cherry-pick the commits and continue the work on the semaphores.

lowlander commented:
> PS: I won't work on this before tomorrow night CEST, feel free to cherry-pick the commits and continue the work on the semaphores.

OK I will give it a try tonight

lowlander commented:
@aurel32 I merged your commits into this PR, probably have to clean things up a bit before upstreaming, but for testing it is OK I think.  It seems to work with my usecase (RX-only, TX-only, and RX+TX, but always just 1 buffer).  I do have the bad feeling it will break when RX and TX are not the same length, where or we have to fill things up with dummy buffers, or some check at startup will need to catch that and give an error.

aurel32 commented:
> @aurel32 I merged your commits into this PR, probably have to clean things up a bit before upstreaming, but for testing it is 
OK I think. It seems to work with my usecase (RX-only, TX-only, and RX+TX, but always just 1 buffer).

Thanks for merging. There is still one issue I have there with your version of the code. You replaced the 

`while (LL_SPI_IsActiveFlag_BSY(spi) == 1 || LL_SPI_IsActiveFlag_TXE(spi) == 0) {`

by

`while (LL_SPI_IsActiveFlag_TXE(spi) == 0) {`

The busy part comes from the original loop. Without it, things are not working as expected here. I do not have a scope at hand, but I guess when stopping the SPI too early, the SS pin is not released correctly (I am using `CONFIG_SPI_STM32_USE_HW_SS=y`). I guess it can also be done using 2 different busy loops.

The bad news however is that the sx1276 driver doesn't work with this PR (but as I said wasn't working before). I am going to debug that.

> I do have the bad feeling it will break when RX and TX are not the same length, where or we have to fill things up with dummy buffers, or some check at startup will need to catch that and give an error.

I agree there is an issue there, but I think it was already the case before that PR. I also thing that not all SPI drivers (even without DMA) handle that correctly, so it's probably just a question of adding a check.

lowlander commented:
> > @aurel32 I merged your commits into this PR, probably have to clean things up a bit before upstreaming, but for testing it is
> > OK I think. It seems to work with my usecase (RX-only, TX-only, and RX+TX, but always just 1 buffer).
> 
> Thanks for merging. There is still one issue I have there with your version of the code. You replaced the
> 
> `while (LL_SPI_IsActiveFlag_BSY(spi) == 1 || LL_SPI_IsActiveFlag_TXE(spi) == 0) {`
> 
> by
> 
> `while (LL_SPI_IsActiveFlag_TXE(spi) == 0) {`
> 
> The busy part comes from the original loop. Without it, things are not working as expected here. I do not have a scope at hand, but I guess when stopping the SPI too early, the SS pin is not released correctly (I am using `CONFIG_SPI_STM32_USE_HW_SS=y`). I guess it can also be done using 2 different busy loops.

I never could get CONFIG_SPI_STM32_USE_HW_SS to work, maybe that was my problem, I will add both again. 

> The bad news however is that the sx1276 driver doesn't work with this PR (but as I said wasn't working before). I am going to debug that.
> 

Do other SPI device work? With other words do you think  it is the SX1276 or your CPU ?


> > I do have the bad feeling it will break when RX and TX are not the same length, where or we have to fill things up with dummy buffers, or some check at startup will need to catch that and give an error.
> 
> I agree there is an issue there, but I think it was already the case before that PR. I also thing that not all SPI drivers (even without DMA) handle that correctly, so it's probably just a question of adding a check.

Easiest way out is probably a check at the start of the transaction to see if tx-len == rx-len when both are > 0 (when rx or tx  is 0 it should be ok, and when both are 0 the transaction makes no sense)

aurel32 commented:
> > The bad news however is that the sx1276 driver doesn't work with this PR (but as I said wasn't working before). I am going to debug that.
> 
> Do other SPI device work? With other words do you think it is the SX1276 or your CPU ?

SPI with DMA works fine when I directly use the spi_read/write/transceive functions. However on the same hardware the Lora driver only works without the DMA. I therefore think it's purely a software issue, there is probably some minor difference.

lowlander commented:
> > > The bad news however is that the sx1276 driver doesn't work with this PR (but as I said wasn't working before). I am going to debug that.
> > 
> > 
> > Do other SPI device work? With other words do you think it is the SX1276 or your CPU ?
> 
> SPI with DMA works fine when I directly use the spi_read/write/transceive functions. However on the same hardware the Lora driver only works without the DMA. I therefore think it's purely a software issue, there is probably some minor difference.

The lora driver seems to do a buffer set with two buffer with RX/TX (same buffer so same size) and a TX only, For the TX only case the dummy RX probably should be the total length of the TX len (same for the other way around) I will check if that is the case later and also fix the BSY waiting.

lowlander commented:
@aurel32 re-added the bsy wait, and changed the length calculation in case only a RX or TX buffer is used, now the dummy DMA transfer should be just as long as the total of the real transfer (not just the size of the first buffer)

aurel32 commented:
So after debugging, it seems that the two main issues are:
- `data->dma_tx.dma_cfg.head_block->dest_address` and `data->dma_rx.dma_cfg.head_block->source_address` are only valid for the first dma transfer. Replacing them by `(uint32_t)LL_SPI_DMA_GetRegAddr(cfg->spi)` fixes the biggest part of the issue.
- Not fully sure, but it sees that we need to call `LL_SPI_DisableDMAReq_TX()` and `LL_SPI_DisableDMAReq_RX()` just before calling `LL_SPI_Disable()` in `transceive_dma()`.

With that the Lora driver seem to work fine, although I only tested it very lightly. I get many `<err> dma_stm32: Unknown interrupt happened.` errors though.

Enough for today, I'll look at that again tomorrow.

lowlander commented:
> So after debugging, it seems that the two main issues are:
> 
> * `data->dma_tx.dma_cfg.head_block->dest_address` and `data->dma_rx.dma_cfg.head_block->source_address` are only valid for the first dma transfer. Replacing them by `(uint32_t)LL_SPI_DMA_GetRegAddr(cfg->spi)` fixes the biggest part of the issue.
> * Not fully sure, but it sees that we need to call `LL_SPI_DisableDMAReq_TX()` and `LL_SPI_DisableDMAReq_RX()` just before calling `LL_SPI_Disable()` in `transceive_dma()`.
> 

Could't we just call LL_SPI_DisableDMAReq_TX and LL_SPI_DisableDMAReq_RX in the ISR just before giving the semaphore ?

lowlander commented:
> So after debugging, it seems that the two main issues are:
> 
> * `data->dma_tx.dma_cfg.head_block->dest_address` and `data->dma_rx.dma_cfg.head_block->source_address` are only valid for the first dma transfer. Replacing them by `(uint32_t)LL_SPI_DMA_GetRegAddr(cfg->spi)` fixes the biggest part of the issue.

data->dma_tx.dma_cfg.head_block gets assigned a pointer to a stack based variable, that looks very wrong.

aurel32 commented:
> Could't we just call LL_SPI_DisableDMAReq_TX and LL_SPI_DisableDMAReq_RX in the ISR just before giving the semaphore ?

Yep, that's a good idea.

aurel32 commented:
> > So after debugging, it seems that the two main issues are:
> > 
> > * `data->dma_tx.dma_cfg.head_block->dest_address` and `data->dma_rx.dma_cfg.head_block->source_address` are only valid for the first dma transfer. Replacing them by `(uint32_t)LL_SPI_DMA_GetRegAddr(cfg->spi)` fixes the biggest part of the issue.
> 
> data->dma_tx.dma_cfg.head_block gets assigned a pointer to a stack based variable, that looks very wrong.

I used the value that was assigned to ` periph_addr` in the old code, but that's indeed plainly wrong. 

> update: I pushed some changes to (hopefully) fix that.

Thanks. It seems to works fine. I still have to do more testing, but this PR looks in a good shape now.

lowlander commented:
> Thanks. It seems to works fine. I still have to do more testing, but this PR looks in a good shape now.

I have the feeling there is still some sort of race condition that prevents the transfer from finishing. Last night my hardware got a watchdog reset and my monitoring software said it was due to a task doing SPI transfers not responding in time.

Apart from finding (which won't be easy, because it ran for atleast 5 hours before it happened) and fixing this I think we need to not K_FOREVER wait on the semaphores, but give them a timeout for when things go really wrong, better fail one transaction than hang the system.

lowlander commented:
@nandojve  I now have a L0 board with a SPI-FeRAM shield and managed to get it working, could you give this PR in combination with  #27723 a try ?

PS: the dts settings I used are;

```
&spi1 {
	status = "okay";

	dmas = <&dma1 3 1 0x28440 0x04>,
	       <&dma1 2 1 0x28480 0x04>;

	dma-names = "tx", "rx";

	cs-gpios = <&gpiob 6 GPIO_ACTIVE_LOW>;
};
```

lowlander commented:
@lochej  could you also give this in combination with  #27723 a try ?

aurel32 commented:
With the recent changes from this PR, the SPI driver works perfectly with DMA on an L433 with a SX1276 chip and the corresponding LoRa driver. I get `<err> dma_stm32: Unknown interrupt happened DMA ISR.` messages, but I have tracked that to a bug in the DMA ISR routine fixed by PR #27723 (and this is not a regression). When using both PRs, all works perfectly without any error message.

The only think that might be missing, is that according to the documentation, the DMA channel must not be configured once enabled. In turns it means that the DMA channel must be disabled at the end of a transfer. It might also bring some power saving, especially for the Lx series. This can be done simply with this patch:

```
        /* end of the transfer : all buffers sent/received */
        LL_SPI_Disable(spi);
 
+       /* Disable DMA channels */
+       dma_stop(data->dev_dma_rx, data->dma_rx.channel);
+       dma_stop(data->dev_dma_tx, data->dma_tx.channel);
+
        spi_stm32_complete(data, spi, ret);
 
        spi_context_release(&data->ctx, ret);
```

nandojve commented:
> @nandojve I now have a L0 board with a SPI-FeRAM shield and managed to get it working, could you give this PR in combination with #27723 a try ?
> 
> PS: the dts settings I used are;
> 
> ```
> &spi1 {
> 	status = "okay";
> 
> 	dmas = <&dma1 3 1 0x28440 0x04>,
> 	       <&dma1 2 1 0x28480 0x04>;
> 
> 	dma-names = "tx", "rx";
> 
> 	cs-gpios = <&gpiob 6 GPIO_ACTIVE_LOW>;
> };
> ```

Hi @lowlander , I'm using AT25 EEPROM driver with an ST M95640 compatible SPI device.
Using same configuration that you shared I got below:

```
uart:~$ eeprom read M95640 1024 8
Reading 8 bytes from EEPROM, offset 1024...
[00:00:01.598,000] <dbg> spi_ll_stm32.spi_stm32_configure: Installed config 0x20000ad4: freq 4000000Hz (div = 8), mode 0/0/0, slave 0
[00:00:01.598,000] <dbg> spi_ll_stm32.spi_context_buffers_setup: tx_bufs 0x200010b0 - rx_bufs 0x200010c0 - 1
[00:00:01.598,000] <dbg> spi_ll_stm32.spi_context_buffers_setup: current_tx 0x200010a8 (1), current_rx 0x200010b8 (1), tx buf/len 0x200010a0/2, rx buf/len 0x200010a4/2
[00:00:01.598,000] <dbg> spi_ll_stm32.spi_context_buffers_setup: tx_bufs 0x200010f8 - rx_bufs 0x20001100 - 1
[00:00:01.598,000] <dbg> spi_ll_stm32.spi_context_buffers_setup: current_tx 0x200010f0 (1), current_rx 0x20001108 (2), tx buf/len 0x200010ec/3, rx buf/len 0x00000000/3
[00:00:01.599,000] <dbg> spi_ll_stm32.spi_context_update_rx: rx buf/len 0x20001160/8
[00:00:02.598,000] <err> os: ***** HARD FAULT *****
[00:00:02.598,000] <err> os: r0/a1:  0x20000fe0  r1/a2:  0x40020000  r2/a3:  0x00000001
[00:00:02.598,000] <err> os: r3/a4:  0x00000001 r12/ip:  0x00000000 r14/lr:  0x00000000
[00:00:02.598,000] <err> os:  xpsr:  0x00000000
[00:00:02.598,000] <err> os: Faulting instruction address (r15/pc): 0x00003e74
[00:00:02.598,000] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
[00:00:02.598,000] <err> os: Current thread: 0x20000468 (logging)
[00:00:02.599,000] <err> os: Halting system
```

When I try use my S2LP native Zephyr driver I got this:
```
D: Initialize S2-LP Transceiver
D: SDN GPIO configured on GPIOA:8
D: IRQ GPIO configured on GPIOA:0 connected at DIO0
D: GPIO configured
D: SPI GPIO CS configured on GPIOB:6
D: SPI configured
D: Installed config 0x200000d0: freq 4000000Hz (div = 8), mode 0/0/0, slave 1
D: tx_bufs 0x20001290 - rx_bufs 0x200012a0 - 1
D: current_tx 0x20001288 (1), current_rx 0x20001298 (1), tx buf/len 0x20001280/2, rx buf/len 0x20001284/2
D: tx_bufs 0x20001280 - rx_bufs 0x20001288 - 1
D: current_tx 0x20001278 (1), current_rx 0x20001290 (2), tx buf/len 0x20001270/2, rx buf/len 0x20001274/2
D: rx buf/len 0x200012b4/2
```

lowlander commented:
> Hi @lowlander , I'm using AT25 EEPROM driver with an ST M95640 compatible SPI device.
> Using same configuration that you shared I got below:
> 
> ```
> uart:~$ eeprom read M95640 1024 8
> Reading 8 bytes from EEPROM, offset 1024...
> [00:00:01.598,000] <dbg> spi_ll_stm32.spi_stm32_configure: Installed config 0x20000ad4: freq 4000000Hz (div = 8), mode 0/0/0, slave 0
> [00:00:01.598,000] <dbg> spi_ll_stm32.spi_context_buffers_setup: tx_bufs 0x200010b0 - rx_bufs 0x200010c0 - 1
> [00:00:01.598,000] <dbg> spi_ll_stm32.spi_context_buffers_setup: current_tx 0x200010a8 (1), current_rx 0x200010b8 (1), tx buf/len 0x200010a0/2, rx buf/len 0x200010a4/2
> [00:00:01.598,000] <dbg> spi_ll_stm32.spi_context_buffers_setup: tx_bufs 0x200010f8 - rx_bufs 0x20001100 - 1
> [00:00:01.598,000] <dbg> spi_ll_stm32.spi_context_buffers_setup: current_tx 0x200010f0 (1), current_rx 0x20001108 (2), tx buf/len 0x200010ec/3, rx buf/len 0x00000000/3
> [00:00:01.599,000] <dbg> spi_ll_stm32.spi_context_update_rx: rx buf/len 0x20001160/8
> [00:00:02.598,000] <err> os: ***** HARD FAULT *****
> [00:00:02.598,000] <err> os: r0/a1:  0x20000fe0  r1/a2:  0x40020000  r2/a3:  0x00000001
> [00:00:02.598,000] <err> os: r3/a4:  0x00000001 r12/ip:  0x00000000 r14/lr:  0x00000000
> [00:00:02.598,000] <err> os:  xpsr:  0x00000000
> [00:00:02.598,000] <err> os: Faulting instruction address (r15/pc): 0x00003e74
> [00:00:02.598,000] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
> [00:00:02.598,000] <err> os: Current thread: 0x20000468 (logging)
> [00:00:02.599,000] <err> os: Halting system
> ```

That looks like some other problem, why would the logging thread crash? Can you try with assert and logging disabled ?

> 
> When I try use my S2LP native Zephyr driver I got this:
> 
> ```
> D: Initialize S2-LP Transceiver
> D: SDN GPIO configured on GPIOA:8
> D: IRQ GPIO configured on GPIOA:0 connected at DIO0
> D: GPIO configured
> D: SPI GPIO CS configured on GPIOB:6
> D: SPI configured
> D: Installed config 0x200000d0: freq 4000000Hz (div = 8), mode 0/0/0, slave 1
> D: tx_bufs 0x20001290 - rx_bufs 0x200012a0 - 1
> D: current_tx 0x20001288 (1), current_rx 0x20001298 (1), tx buf/len 0x20001280/2, rx buf/len 0x20001284/2
> D: tx_bufs 0x20001280 - rx_bufs 0x20001288 - 1
> D: current_tx 0x20001278 (1), current_rx 0x20001290 (2), tx buf/len 0x20001270/2, rx buf/len 0x20001274/2
> D: rx buf/len 0x200012b4/2
> ```

Could you share your software with me? I have a STEBAL-FKI868V2 kit here now, so I could try to reproduce your problem.

lowlander commented:
> The only think that might be missing, is that according to the documentation, the DMA channel must not be configured once enabled. In turns it means that the DMA channel must be disabled at the end of a transfer. It might also bring some power saving, especially for the Lx series. This can be done simply with this patch:
> 
> ```
>         /* end of the transfer : all buffers sent/received */
>         LL_SPI_Disable(spi);
>  
> +       /* Disable DMA channels */
> +       dma_stop(data->dev_dma_rx, data->dma_rx.channel);
> +       dma_stop(data->dev_dma_tx, data->dma_tx.channel);
> +
>         spi_stm32_complete(data, spi, ret);
>  
>         spi_context_release(&data->ctx, ret);
> ```

Will add that and a check to see if total-tx-len and total-rx-len are the same (when both of them are != 0) 
I would also like to reproduce @nandojve problem, to see it that exposes any fundamental design errors.

nandojve commented:
> > Hi @lowlander , I'm using AT25 EEPROM driver with an ST M95640 compatible SPI device.
> > Using same configuration that you shared I got below:
> > ```
> > uart:~$ eeprom read M95640 1024 8
> > Reading 8 bytes from EEPROM, offset 1024...
> > [00:00:01.598,000] <dbg> spi_ll_stm32.spi_stm32_configure: Installed config 0x20000ad4: freq 4000000Hz (div = 8), mode 0/0/0, slave 0
> > [00:00:01.598,000] <dbg> spi_ll_stm32.spi_context_buffers_setup: tx_bufs 0x200010b0 - rx_bufs 0x200010c0 - 1
> > [00:00:01.598,000] <dbg> spi_ll_stm32.spi_context_buffers_setup: current_tx 0x200010a8 (1), current_rx 0x200010b8 (1), tx buf/len 0x200010a0/2, rx buf/len 0x200010a4/2
> > [00:00:01.598,000] <dbg> spi_ll_stm32.spi_context_buffers_setup: tx_bufs 0x200010f8 - rx_bufs 0x20001100 - 1
> > [00:00:01.598,000] <dbg> spi_ll_stm32.spi_context_buffers_setup: current_tx 0x200010f0 (1), current_rx 0x20001108 (2), tx buf/len 0x200010ec/3, rx buf/len 0x00000000/3
> > [00:00:01.599,000] <dbg> spi_ll_stm32.spi_context_update_rx: rx buf/len 0x20001160/8
> > [00:00:02.598,000] <err> os: ***** HARD FAULT *****
> > [00:00:02.598,000] <err> os: r0/a1:  0x20000fe0  r1/a2:  0x40020000  r2/a3:  0x00000001
> > [00:00:02.598,000] <err> os: r3/a4:  0x00000001 r12/ip:  0x00000000 r14/lr:  0x00000000
> > [00:00:02.598,000] <err> os:  xpsr:  0x00000000
> > [00:00:02.598,000] <err> os: Faulting instruction address (r15/pc): 0x00003e74
> > [00:00:02.598,000] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
> > [00:00:02.598,000] <err> os: Current thread: 0x20000468 (logging)
> > [00:00:02.599,000] <err> os: Halting system
> > ```
> 
> That looks like some other problem, why would the logging thread crash? Can you try with assert and logging disabled ?
> 
> > When I try use my S2LP native Zephyr driver I got this:
> > ```
> > D: Initialize S2-LP Transceiver
> > D: SDN GPIO configured on GPIOA:8
> > D: IRQ GPIO configured on GPIOA:0 connected at DIO0
> > D: GPIO configured
> > D: SPI GPIO CS configured on GPIOB:6
> > D: SPI configured
> > D: Installed config 0x200000d0: freq 4000000Hz (div = 8), mode 0/0/0, slave 1
> > D: tx_bufs 0x20001290 - rx_bufs 0x200012a0 - 1
> > D: current_tx 0x20001288 (1), current_rx 0x20001298 (1), tx buf/len 0x20001280/2, rx buf/len 0x20001284/2
> > D: tx_bufs 0x20001280 - rx_bufs 0x20001288 - 1
> > D: current_tx 0x20001278 (1), current_rx 0x20001290 (2), tx buf/len 0x20001270/2, rx buf/len 0x20001274/2
> > D: rx buf/len 0x200012b4/2
> > ```
> 
> Could you share your software with me? I have a STEBAL-FKI868V2 kit here now, so I could try to reproduce your problem.

Sure, https://github.com/nandojve/zephyr/tree/ht-micron_s2-lp
S2LP have a BUG. If you try use another device on SPI bus you need put S2LP on shutdown otherwise you will see only zeros on the bus. STEVAL-FKI868V2 need be reworked to use both S2LP and EEPROM on regular Arduino pins, need move 3 resistors I think. LED gpio is the same SCK clock, be carefully. L053 have internal EEPROM, make sure you are using AT25 driver.
To use S2LP you need build samples/drivers/sigfox/send

FRASTM commented:
with this change, can 2 DMA instances be enabled with a loop transfer transfer (tests/drivers/dma/loop_transfer) on the 1st one.

lowlander commented:
@nandojve could you try with disabling logging CONFIG_LOG=n and disabling asserts CONFIG_ASSERT=n ? I think I have found a race condition, where the reloading of the RX buffer isn't fast enough so it misses bytes at the start, and because of that never finishes because it still wants to receive the set number of bytes.

lowlander commented:
@aurel32 could you try with logging and assert enabled and/or higher bus speed (I use 5MHz, when I reduce it to 500kHz it works fine again) There seems the be a timing problem when using multiple buffers in transceive mode. For example reading from SPI FRAM, there are two buffers one 3 byte command buffer and the buffer that should get the read data. When the TX is done with the 3 bytes, it will get the IRQ and load the next buffer, so than it directly starts sending again (and with that receiving) but the RX done IRQ comes after that and so the loading of the next RX buffer is not yet done. If there now is the slightest delay between the TX-done and RX-done IRQ (for example the 50 micro seconds for assert checking) we loose bytes (and with that lock up). And this was with 5MHz SPI, if that would run at 50MHz there is simply no way to be on time.

So I think we have to move the reload out of the ISR again and make sure the rx and tx buffer-sets are always exactly the same count and  size. Or make sure the buffers given to the DMA are the same size, so we have to split up the buffers to always match, if the TX-set has 2 buffers of 2 and 4 bytes and the rx set 2 buffers of 3 bytes, we will have to split them in 3 buffers of 2, 1, and 3 bytes, so the RX and TX DMA are always finished after the same number of bytes, and so the reloading happens at the same time (when both RX and TX are disabled).

aurel32 commented:
> @aurel32 could you try with logging and assert enabled and/or higher bus speed (I use 5MHz, when I reduce it to 500kHz it works fine again) There seems the be a timing problem when using multiple buffers in transceive mode. For example reading from SPI FRAM, there are two buffers one 3 byte command buffer and the buffer that should get the read data. When the TX is done with the 3 bytes, it will get the IRQ and load the next buffer, so than it directly starts sending again (and with that receiving) but the RX done IRQ comes after that and so the loading of the next RX buffer is not yet done. If there now is the slightest delay between the TX-done and RX-done IRQ (for example the 50 micro seconds for assert checking) we loose bytes (and with that lock up). And this was with 5MHz SPI, if that would run at 50MHz there is simply no way to be on time.

I have tried that, but I have not been able to reproduce the issue. My SPI runs at 10MHz which is the maximum supported by the SX1276. OTOH, it's on a L4 CPU, I guess there is way more power than on your L0 system.

> So I think we have to move the reload out of the ISR again and make sure the rx and tx buffer-sets are always exactly the same count and size. Or make sure the buffers given to the DMA are the same size, so we have to split up the buffers to always match, if the TX-set has 2 buffers of 2 and 4 bytes and the rx set 2 buffers of 3 bytes, we will have to split them in 3 buffers of 2, 1, and 3 bytes, so the RX and TX DMA are always finished after the same number of bytes, and so the reloading happens at the same time (when both RX and TX are disabled).

I have seen that you made that change. It still works fine for me. I guess in a future we can use the same principle that you use in the ISR to avoid a context change, which is probably more expensive than reloading the DMA. But it's probably better to do that in another PR to avoid waiting indefinitely for it to be ready.

That part looks good to me, it just lacks the dma_stop at the end of the transfer. I have more comments on the second commit concerning the cache handling. Maybe it's also better to do that in a different PR?

aurel32 commented on drivers/spi/Kconfig.stm32:
```diff
@@ -24,6 +24,19 @@ config SPI_STM32_DMA
 	  Enable the SPI DMA mode for SPI instances
 	  that enable dma channels in their device tree node.
 
+config SPI_STM32_UNALIGNED_DMA
```
Most of the time the SPI driver will be used by another driver, so users can't control if the buffer is aligned or not. I therefore believe this should be enabled by default.

aurel32 commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -53,10 +56,36 @@ LOG_MODULE_REGISTER(spi_ll_stm32);
+#define SPI_STM32_ROUND_DOWN(x) \
+			ROUND_DOWN((uint32_t)(x), __SCB_DCACHE_LINE_SIZE)
+
+void spi_stm32_flush_cache(const void *ptr, uint32_t size)
```
In practice both `spi_stm32_flush_cache` and `spi_stm32_invalidate_cache` are called with aligned address and length that are a multiple of `__SCB_DCACHE_LINE_SIZE`, so there is no need for all that alignment around.

lowlander replied:
Yeah those checks were from before I realized I could not just invalidate a few bytes by finding the cacheline and invalidating the that line. The rounding is done again in the actual ARM code, so these checks could go.  Maybe a ASSERT could be put in place to make it very clear the address and size must be aligned ?

aurel32 replied:
The ASSERT is a really good idea.

aurel32 commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -77,6 +106,19 @@ static void dma_callback(struct device *dev, void *arg,
 		} else if (channel == data->dma_rx.channel) {
 			/* this part of the transfer ends */
 			data->status_flags |= SPI_STM32_DMA_RX_DONE_FLAG;
+#if CONFIG_SPI_STM32_UNALIGNED_DMA
```
Why is the cache invalidation done only when `CONFIG_SPI_STM32_UNALIGNED_DMA` is enabled? The cache also needs to be invalidated in case the buffers are assumed to be aligned.

lowlander replied:
The Idea was that if the user does not enable the CONFIG_SPI_STM32_UNALIGNED_DMA option he is responsible for cache management, or by invalidating his buffers after SPI is done or by using uncachable memory. Because the driver can not know if an aligned buffer of for example 3 bytes can be invalidated, it could only invalidate buffers that are exactly correct aligned and have a multiple of the cache line size, which will give most certainly weird bug reports, like; "when I send 32 bytes it works, but when I send 48 half my data is not there"

aurel32 replied:
This is not clear from the description of the option, which seems to imply that things should just work if the buffer are aligned. 
Also I am not fully sure, but I think that you can actually need to use an extra buffer for TX, as long as all the cache lines containing the data are flushed. The extra cache flushes should just make a small performances impact. For RX and cache invalidation, the buffer must definitely be aligned on the cache line addresses and size to avoid data losses (or just ensure that no other data is on the affected cache lines, which is basically the same).

lowlander replied:
The idea was that a cache flushes could over write memory that was written by another RX DMA, but when RX DMA is always aligned that can never happen. Even if they use the same buffer the flush would happen before the RX DMA starts. So the TX buffer could be removed.

On the other hand when one would also want to support using non-DMA-able memory the write buffer must be copied to memory that is DMA-able first.

Like you mentioned the Kconfig name is confusing and the whole thing will need way more discussion, I really think it is better to move it to its own PR.

aurel32 commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -209,25 +265,74 @@ static int spi_dma_move_buffers(struct device *dev, size_t len)
+#if CONFIG_SPI_STM32_UNALIGNED_DMA
+
+	if (rx_buf != NULL) {
+		uint32_t addr_diff = SPI_STM32_ROUND_UP(rx_buf)
```
The name might be misleading, it's not exactly an address difference, but rather a length up to the next aligned boundary. This differs from an address difference by 1.

lowlander replied:
yeah well naming things is the most difficult part of programming.

aurel32 commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -669,83 +674,77 @@ static int transceive_dma(struct device *dev,
+		} else if (data->ctx.tx_len == 0) {
+			dma_len = data->ctx.rx_len;
+		} else {
+			dma_len = data->ctx.tx_len < data->ctx.rx_len ?
```
please use MIN instead.

lowlander commented:
> I have seen that you made that change. It still works fine for me. I guess in a future we can use the same principle that you use in the ISR to avoid a context change, which is probably more expensive than reloading the DMA. But it's probably better to do that in another PR to avoid waiting indefinitely for it to be ready.
> 

Not might be possible to do it all in the ISR, might get a bit tricky because the next transfer can only be started after both RX and TX are done. For the moment I think it would be a good thing to first at least have something that works. 

> That part looks good to me, it just lacks the dma_stop at the end of the transfer. 

I will put that back in, that is no problem.

> I have more comments on the second commit concerning the cache handling. Maybe it's also better to do that in a different PR?

I will split that out to a new PR, for someone else to finish, F7 is not one of my platforms I need and I already spend way to mush time on this. And I have the feeling cache handling is still not that well supported in Zephyr and could become a lengthy discussion.

lowlander commented:
Moved the F7 cache handling code to its own PR #27911

lowlander committed:
drivers: dma: stm32: change order of disabling and checking result

The stm32_dma_disable_stream first checked if the stream was
disabled and if so returned OK. If it wasn't disabled it
tried to disable it and returned -EAGAIN.

The function is used in loops that try to disable the stream
by calling this function and if it fails wait for 1ms and
retry.

Becuase this function the first time (if the stream wasn't
disabled already) fails there is always a 1 ms delay. For
the SPI driver, that has a RX and TX stream, this means
a 2ms delay between the last data and CS going high.

By first trying to disable the stream and than checking
if it succeded most of the time the first call disables
the stream and the 1ms delay isn't needed.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander commented:
@lochej @FRASTM @erwango @cybertale @aurel32 Now that the DMA changes are in master this PR should be ready to be merged if there are no more change requests. @cybertale could you take a look at the small DMA code change, that I only noticed when measuring the SPI signal with a scope. I think the change is OK, but a second look by someone else never hurts.

tbursztyka commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -53,186 +53,176 @@ LOG_MODULE_REGISTER(spi_ll_stm32);
-	if (data->dma_rx.src_addr_increment) {
-		blk_cfg.source_addr_adj = DMA_ADDR_ADJ_INCREMENT;
+	if (buf == NULL) {
+		/* if rx buff is null, then write data to dummy address. */
```
Nice you could fix that

lowlander replied:
not sure what you exactly mean ? The comment ? or the code ?

tbursztyka replied:
the fact that rx buf can be NULL and you handle that case, afaik that was not working before with DMA enabled.

lowlander replied:
Oh I read that wrong I thought "Nice if you could fix that" :-)

erwango commented:
@lowlander I'm tagging this PR with V2.4.0 milestone.
V2.4.0 merge window is closing today, but we'll still be able to merge bug fixes as long as the bug relates to a filled issue (with a "fixes #foobar" reference).
So, if not merged today, an issue will need to be filled.

lowlander commented:
> @lowlander I'm tagging this PR with V2.4.0 milestone.
> V2.4.0 merge window is closing today, but we'll still be able to merge bug fixes as long as the bug relates to a filled issue (with a "fixes #foobar" reference).
> So, if not merged today, an issue will need to be filled.

@erwango there might actually already be some issues that could have been fixed by this PR, might have to go trough SPI related issues that are out standing to see if the reporters can test with this PR

aurel32 reviewed (approved):
The latest changes look good to me. I have just re-tested this PR on my setup, and things still work as expected. Thanks for your work!
